route/bgmetadata.go

```diff
@@ -0,0 +1,153 @@
+type BgMetadata struct {
+	baseRoute
+	shards []shard
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	bfCfg  BloomFilterConfig
+}
+
+// NewBgMetadataRoute creates BgMetadata, starts sharding and filtering incoming metrics.
+// Runs a goroutines for each shard to handle incoming metrics and periodic cleanup of bloom filters
+func NewBgMetadataRoute(key, prefix, sub, regex string, bfCfg *BloomFilterConfig) (*BgMetadata, error) {
+	m := BgMetadata{
+		baseRoute: *newBaseRoute(key, "bg_metadata"),
+		shards:    make([]shard, bfCfg.ShardingFactor),
+		bfCfg:     *bfCfg,
+	}
+
+	if bfCfg.ClearWait != 0 {
+		m.bfCfg.ClearWait = bfCfg.ClearWait
+	} else {
+		m.bfCfg.ClearWait = bfCfg.ClearInterval / time.Duration(bfCfg.ShardingFactor)
+		m.logger.Info(fmt.Sprintf("setting clear_wait value to %s", m.bfCfg.ClearWait.String()))
+	}
+
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// init every shard with filter and channel
+	for shardNum := 0; shardNum < bfCfg.ShardingFactor; shardNum++ {
+		m.shards[shardNum] = shard{
+			filter:  *bloom.NewWithEstimates(bfCfg.N, bfCfg.P),
+			channel: make(chan []byte, 500),
+		}
+		go m.handleMetric(shardNum)
+	}
+	m.logger.Info("all metric handlers started")
+
+	go m.clearBloomFilter()
+
+	// matcher required to initialise route.Config for routing table, othewise it will panic
+	mt, err := matcher.New(prefix, sub, regex)
+	if err != nil {
+		return nil, err
+	}
+	m.config.Store(baseConfig{*mt, nil})
+
+	return &m, nil
+}
+
+func (m *BgMetadata) handleMetric(shardNum int) {
+	m.wg.Add(1)
+	defer m.wg.Done()
+	m.logger.Debug(fmt.Sprintf("starting metric handler for shard %d", shardNum+1))
+	for {
+		select {
+		case <-m.ctx.Done():
+			close(m.shards[shardNum].channel)
+			return
+		case metric := <-m.shards[shardNum].channel:
+			if !m.shards[shardNum].filter.Test(metric) {
+				m.shards[shardNum].filter.Add(metric)
+				// increase outgoing metric prometheus counter
+				m.rm.OutMetrics.Inc()
+				// do nothing for now
+				m.logger.Debug(fmt.Sprintf("adding metric to metadata: %s", string(metric)))
+			} else {
+				// don't output metrics already in the filter
+				m.logger.Debug(fmt.Sprintf("skipping metric: %s", string(metric)))
+			}
+		}
+	}
+}
+
+func (m *BgMetadata) clearBloomFilter() {
+	m.logger.Info("starting bloom filter clear goroutine")
+	m.wg.Add(1)
+	defer m.wg.Done()
+	t := time.NewTicker(m.bfCfg.ClearInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+		case <-t.C:
+			for i := range m.shards {
+				m.shards[i].lock.Lock()
+				m.shards[i].filter.ClearAll()
+				m.logger.Debug(fmt.Sprintf("clearing all filters for shard %d", i+1))
+				m.shards[i].lock.Unlock()
+				time.Sleep(m.bfCfg.ClearWait)
+			}
+		}
+	}
+}
+
+func getShardNum(metricName string, shardingFactor int) uint32 {
+	return crc32.Checksum([]byte(metricName), crc32.MakeTable(crc32.IEEE)) % uint32(shardingFactor)
```

Comment:
why not using the ChecksumIEEE directly instead of creating a new table each time

Reply:
I missed it. Replaced.